<a href="https://colab.research.google.com/github/dazzens/dazzens/blob/main/EDA_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

------------------
---------
-----

# **Пет проект EDA (исследовательский анализ данных).**




Автор: А.Г.Х.

Начало: 22.06.2026


Окончание: --.--.2026

----



##**Бизнес-контекст**

- Интернет-магазин продает товары для дома и декора по всему миру. За последний год ассортимент вырос на 30%, но темпы роста выручки замедлились. Необходимо исследовать данные о продажах, чтобы понять, какие товары и страны приносят основную прибыль, кто самые ценные клиенты и как оптимизировать ценовую политику.

-----

# **Цель исследования**

- Выявить ключевые факторы, влияющие на динамику продаж интернет-магазина, для разработки рекомендаций по увеличению выручки и удержанию клиентов.


##**Вопросы:**


- 1. Как менялись продажи по месяцам? Динамика выручки, количества заказов и сезонность.


- 2. Какие товары самые популярные? Топ по количеству продаж, топ по выручке, топ по частоте заказов.

- 3. Какой средний чек и размер корзины? Среднее количество товаров в заказе, средняя сумма заказа

- 4. Какие страны приносят больше всего выручки? Топ по выручке, топ по количеству заказов, топ по среднему чеку

- 5. Кто самые прибыльные клиенты? Топ клиентов по сумме трат, по частоте покупок, анализ повторных покупок

- 6. Как цена влияет на продажи? Распределение цен, зависимость количества продаж от цены, оптимальные ценовые сегменты

- 7. Какие товары, страны и клиентские сегменты вносят наибольший вклад в выручку?


## **Содержание проекта**



1.  **Первичный осмотр и подготовка данных**
    *   Импорт библиотек и загрузка данных.
    *   Предварительный анализ структуры данных (типы, объем, пропуски).
    *   Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
    *   Проверка и обработка пропусков (`CustomerNo`).
    *   Проверка дубликатов и выбросов (`Price`, `Quantity`).

2.  **Предварительный анализ и создание новых признаков**
    *   Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
    *   Создание признаков: месяц, день недели, год из даты.
    *   Создание признака: общая сумма заказа (`Price` x `Quantity`).
    *   Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

3.  **Исследовательский анализ данных (EDA)**
    *   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
    *   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
    *   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
    *   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
    *   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
    *   Анализ ценовых сегментов: зависимость количества продаж от цены.

4.  **Ключевые выводы и рекомендации**
    *   Формирование основных инсайтов по каждому блоку анализа.
    *   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.


    

## **Описание данных**

**Транзакционные данные (онлайн-магазин):**

| | | |
|------|----------|------------|
| `TransactionNo` | Уникальный идентификатор транзакции/заказа | `object` |
| `Date` | Дата совершения покупки | `object` |
| `ProductNo` | Уникальный идентификатор товара | `object` |
| `ProductName` | Название товара | `object` |
| `Price` | Цена за единицу товара | `float64` |
| `Quantity` | Количество единиц товара в заказе | `int64` |
| `CustomerNo` | Уникальный идентификатор клиента | `float64` |
| `Country` | Страна клиента | `object` |

---

**Характеристики данных:**

|  |  |
|------------|----------|
| **Объем данных** | 536 350 записей |
| **Количество клиентов** | 4 738 |
| **Количество уникальных товаров** | 3 768 |
| **Количество стран** | 38 |
| **Пропуски** | 55 записей без `CustomerNo` (0.01%) |



----
----

# **Этап 1 Первичный осмотр и подготовка данных**

- Импорт библиотек и загрузка данных.
- Предварительный анализ структуры данных (типы, объем, пропуски).
- Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
- Проверка и обработка пропусков (`CustomerNo`).
- Проверка дубликатов и выбросов (`Price`, `Quantity`).



## **Шаг 1. Загрузка данных**

- Импорт библиотек и загрузка датасета из внешнего источника.

---
### **Библиотеки**




In [40]:
#загруза библиотек

try:
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    import kagglehub
    import os
    print("Все библиотеки загружены")

except ImportError:
    !pip install numpy pandas matplotlib seaborn   kagglehub -q
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    import kagglehub
    import os
    print("Библиотеки загружены через pip")



Все библиотеки загружены


### **Загрузка датасета**

In [41]:
#Загрузка датасета
path = kagglehub.dataset_download("gabrielramos87/an-online-shop-business")
file_path = os.path.join(path, [x for x in os.listdir(path) if x.endswith('.csv')][0])

#Сохранение
df = pd.read_csv(file_path)



Using Colab cache for faster access to the 'an-online-shop-business' dataset.


**Итого:**

 - Установлены необходимые библиотеки и датасет сохранён в `df`

-----
----
----

## **Шаг 2. Общий осмотр данных**

Первичный анализ структуры данных включает оценку следующих параметров, таких как:

- размерность датасета
- типы признаков и их соответствие содержанию
- наличие пропусков и их доля в каждом поле
- основные статистические характеристики числовых переменных

In [42]:
# Основная информация
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 536350 entries, 0 to 536349
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  536350 non-null  object 
 1   Date           536350 non-null  object 
 2   ProductNo      536350 non-null  object 
 3   ProductName    536350 non-null  object 
 4   Price          536350 non-null  float64
 5   Quantity       536350 non-null  int64  
 6   CustomerNo     536295 non-null  float64
 7   Country        536350 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.7+ MB


In [43]:
#Вывод первых 10 строк
df.head(10)


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
0,581482,12/9/2019,22485,Set Of 2 Wooden Market Crates,21.47,12,17490.0,United Kingdom
1,581475,12/9/2019,22596,Christmas Star Wish List Chalkboard,10.65,36,13069.0,United Kingdom
2,581475,12/9/2019,23235,Storage Tin Vintage Leaf,11.53,12,13069.0,United Kingdom
3,581475,12/9/2019,23272,Tree T-Light Holder Willie Winkie,10.65,12,13069.0,United Kingdom
4,581475,12/9/2019,23239,Set Of 4 Knick Knack Tins Poppies,11.94,6,13069.0,United Kingdom
5,581475,12/9/2019,21705,Bag 500g Swirly Marbles,10.65,24,13069.0,United Kingdom
6,581475,12/9/2019,22118,Joy Wooden Block Letters,11.53,18,13069.0,United Kingdom
7,581475,12/9/2019,22119,Peace Wooden Block Letters,12.25,12,13069.0,United Kingdom
8,581475,12/9/2019,22217,T-Light Holder Hanging Lace,10.65,12,13069.0,United Kingdom
9,581475,12/9/2019,22216,T-Light Holder White Lace,10.55,24,13069.0,United Kingdom


In [44]:
#Осмотр случайных строк
df.sample(10)


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
74363,576190,11/14/2019,22354,Retrospot Padded Seat Cushion,14.09,4,15039.0,United Kingdom
59941,577162,11/18/2019,23355,Hot Water Bottle Keep Calm,15.32,8,13323.0,United Kingdom
451822,543399,2/8/2019,85186A,Easter Bunny Garland Of Flowers,11.10,1,13399.0,United Kingdom
172051,568868,9/29/2019,23404,Home Sweet Home Blackboard,15.32,6,12957.0,United Kingdom
458862,542893,2/1/2019,22772,Pink Drawer Knob Acrylic Edwardian,11.53,12,18202.0,United Kingdom
452324,543466,2/8/2019,22855,Fine Wicker Heart,11.53,12,16713.0,United Kingdom
491041,540372,1/6/2019,22466,Fairy Tale Cottage Night Light,12.25,12,13081.0,United Kingdom
14524,580667,12/5/2019,21906,Pharmacie First Aid Tin,7.24,2,15031.0,United Kingdom
402961,547866,3/27/2019,22997,Travel Card Wallet Union Jack,10.68,24,14878.0,United Kingdom
482850,540977,1/12/2019,35651,Vintage Bead Pink Scarf,13.69,1,14977.0,United Kingdom


In [45]:
#Осмотр по различным показателям
df.describe()


,Price,Quantity,CustomerNo
count,536350.000000,536350.000000,536295.000000
mean,12.662182,9.919347,15227.893178
std,8.490450,216.662300,1716.582932
min,5.130000,-80995.000000,12004.000000
25%,10.990000,1.000000,13807.000000
50%,11.940000,3.000000,15152.000000
75%,14.090000,10.000000,16729.000000
max,660.620000,80995.000000,18287.000000


### **Результаты первичного осмотра**

**Общая информация:**
- Датасет содержит 536 350 записей, 8 столбцов.
- Пропуски: в поле `CustomerNo`.
- Типы данных: `Date` и `CustomerNo` требуют преобразования для дальнейшего анализа.
- `Price`: среднее — 12.66 руб., медиана — 11.94 руб. Диапазон: от 5.13 до 660.62 руб.
- `Quantity`: среднее — 9.92 шт., медиана — 3 шт. Диапазон: от -80 995 до 80 995 шт., что указывает на наличие экстремальных значений.
  
**Выводы:**
  
  -  Данные содержат аномалии в `Quantity` (отрицательные и экстремально высокие значения) и выбросы в `Price`.
  - Требуется преобразование типов и принятие решения по обработке аномалий. В остальном данные пригодны для анализа.
  - Требуется осомотр пропусков в поле `CustomerNo`.


## **Шаг 3. Обработка типов данных**

- проверка и приведение типов данных к форматам, соответствующим их содержанию
- обеспечение согласованности типов для последующих операций
- оптимизация структуры данных для дальнейшей работы
- `Date` преобразовать в `datetime` для временной аналитики
- `CustomerNo` привести  к `Int64` для корректной группировки клиентов

In [46]:
#Осмотр типов данных

df.dtypes

,0
TransactionNo,object
Date,object
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,float64
Country,object


In [47]:
# преобразование данных
df['Date'] = pd.to_datetime(df['Date'])
df['CustomerNo'] = df['CustomerNo'].astype('Int64')

#проверка
df.dtypes

,0
TransactionNo,object
Date,datetime64[ns]
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,Int64
Country,object


### **Итог преобразования**

Типы данных приведены к форматам, необходимым для дальнейшего анализа

- `Date` преобразована в `datetime`, что открывает доступ к исследованию временных рядов и сезонных срезов;
- `CustomerNo` переведён в `Int64` с поддержкой пропусков, что обеспечивает корректную идентификацию клиентов при группировках.

Остальные типы данных оставлены без изменений.

---
----

## **Шаг 4 Поиск и обработка пропусков**

- Проверка данных на наличие пропущенных значений
- оценка их объема
- анализ возможного влияния на результаты исследования и принятие
- решения по их обработке

In [48]:
# колличесвто и доля пропусков

df.isnull().sum(), df.isnull().mean()

(TransactionNo     0
 Date              0
 ProductNo         0
 ProductName       0
 Price             0
 Quantity          0
 CustomerNo       55
 Country           0
 dtype: int64,
 TransactionNo    0.000000
 Date             0.000000
 ProductNo        0.000000
 ProductName      0.000000
 Price            0.000000
 Quantity         0.000000
 CustomerNo       0.000103
 Country          0.000000
 dtype: float64)

In [49]:
# отображение данных

df[df['CustomerNo'].isnull()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,<NA>,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,<NA>,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,<NA>,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,<NA>,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,<NA>,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,<NA>,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,<NA>,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,<NA>,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,<NA>,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,<NA>,United Kingdom


####**Описание и дальнейшие действия:**

* Пропуски обнаружены только в поле `CustomerNo` и составляют **55** записей (0.01%).
* Для сохранения всех наблюдений пропуски будут заменены значением `-1`, которое в дальнейшем будет обозначать неизвестного клиента

----

In [50]:
# Замена NaN на специальное значение
df['CustomerNo'] = df['CustomerNo'].fillna(-1)

# проверка после заполнения пропусков
display(df[df['CustomerNo'] == -1].head(10))

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,-1,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,-1,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,-1,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,-1,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,-1,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,-1,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,-1,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,-1,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,-1,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,-1,United Kingdom


### **Итог работы c пропусками:**


Пропуски в `CustomerNo` обработаны заменой на значение `-1`.


Пропущенных значений в датасете не осталось.



---
---

## **Шаг 5. Поиск и анализ  явных/неявных дубликатов**

- Выявление избыточных записей (полностью совпадающих строк)

- оценка их объёма и доли в общем массиве данных

- анализ возможного влияния дубликатов на агрегированные метрики

- принятие решения по обработке: удаление или агрегация

---

###**Анализ явных дубликатов:**

- поиск полностью совпадающих строк по всем колонкам
- оценка количества и доли дубликатов в данных
- анализ влияния на метрики (выручка, количество заказов, товаров)


In [51]:
# Количество и доля явных дубликатов

duplicates_count = df.duplicated().sum()
duplicates_percent = (duplicates_count / len(df)) * 100

print(f"Количество дубликатов: {duplicates_count}")
print(f"Процент дубликатов: {duplicates_percent:.2f}%")

Количество дубликатов: 5200
Процент дубликатов: 0.97%


In [52]:
#Вывод записей дубликатов

df[df.duplicated()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
985,581497,2019-12-09,21481,Fawn Blue Hot Water Bottle,7.24,1,17497,United Kingdom
1365,581538,2019-12-09,23275,Set Of 3 Hanging Owls Ollie Beak,6.19,1,14446,United Kingdom
1401,581538,2019-12-09,22992,Revolver Wooden Ruler,6.19,1,14446,United Kingdom
1406,581538,2019-12-09,22694,Wicker Star,6.19,1,14446,United Kingdom
1409,581538,2019-12-09,23343,Jumbo Bag Vintage Christmas,6.19,1,14446,United Kingdom
3255,581237,2019-12-08,23498,Classic Bicycle Clips,6.19,1,15443,United Kingdom
3259,581237,2019-12-08,23497,Classic Chrome Bicycle Bell,6.19,1,15443,United Kingdom
3261,581237,2019-12-08,23509,Mini Playing Cards Fun Fair,6.19,2,15443,United Kingdom
3284,581237,2019-12-08,23378,Pack Of 12 50'S Christmas Tissues,6.19,1,15443,United Kingdom
3593,581253,2019-12-08,23427,Stool Home Sweet Home,6.19,1,16891,United Kingdom


####**Дальнейшие действия с явными дубликатами:**


Обнаружено **5200** дубликатов (0.97% от всех записей).  

Удаление не повлияет на общую структуру данных и исключит искажение метрик.

In [53]:
#Удаление дубликатов и проверка

df = df.drop_duplicates()

df.duplicated().sum()

np.int64(0)

####**Итоги обработки:**

Явные дубликаты полностью удалены.

 В датасете не осталось полностью совпадающих строк.


----
-----

### **Анализ неявных дубликатов:**

- Проверка наличия повторяющихся товаров в рамках одной транзакции.
- Оценка доли неявных дубликатов.
- Анализ структуры повторяющихся записей.
- Проверка возможности их объединения без потери информации.
- Выбор и обоснование обработки.

In [54]:
# Поиск неявных дубликатов

implicit_duplicates = (
    df.groupby(['TransactionNo', 'ProductNo'])
      .size()
      .reset_index(name='count')
      .query('count > 1')
)

display(implicit_duplicates.head(10))

print(f'Найдено комбинаций с повторяющимися товарами: {len(implicit_duplicates)}')



,TransactionNo,ProductNo,count
133,536381,71270,2
515,536409,85116,2
518,536409,90199C,3
529,536412,21448,2
535,536412,21738,3
538,536412,22077,3
541,536412,22144,2
545,536412,22243,3
547,536412,22273,2
562,536412,22749,2


Найдено комбинаций с повторяющимися товарами: 5553


In [55]:
# Строки, относящиеся к найденным комбинациям

duplicate_rows = df.merge(
    implicit_duplicates[['TransactionNo', 'ProductNo']],
    on=['TransactionNo', 'ProductNo']
)

print(f'Количество строк с повторами: {len(duplicate_rows)}')
print(f'Доля строк от общего объема данных: {len(duplicate_rows) / len(df) * 100:.2f}%')

Количество строк с повторами: 11336
Доля строк от общего объема данных: 2.13%


In [56]:
# Пример одной транзакции с повторяющимися товарами

sample_transaction = implicit_duplicates['TransactionNo'].iat[0]

display(
    df.loc[
        df['TransactionNo'] == sample_transaction,
        ['TransactionNo', 'Date', 'ProductNo', 'ProductName', 'Price', 'Quantity']
    ].sort_values('ProductNo')
)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity
533393,536381,2018-12-01,15056N,Edwardian Parasol Natural,16.35,2
533379,536381,2018-12-01,21166,Cook With Wine Metal Sign,12.25,1
533378,536381,2018-12-01,21169,You're Confusing Me Metal Sign,11.98,3
533380,536381,2018-12-01,21175,Gin And Tonic Diet Metal Sign,12.40,2
533400,536381,2018-12-01,21523,Doormat Fancy Font Home Sweet Home,17.17,10
533391,536381,2018-12-01,21533,Retrospot Large Milk Jug,15.32,1
533392,536381,2018-12-01,21557,Set Of 6 Funky Beakers,13.27,2
533371,536381,2018-12-01,21672,White Spot Red Ceramic Drawer Knob,11.53,6
533397,536381,2018-12-01,21731,Red Toadstool Led Night Light,11.94,2
533377,536381,2018-12-01,21934,Skull Shoulder Bag,11.94,10


---
Также, для точности, необходимо будет:

- проверить корректность объединения данных

- убедиться, что повторяющиеся товары не имеют разницу по цене

In [57]:
price_check = (
    duplicate_rows
    .groupby(['TransactionNo', 'ProductNo'])
    .agg(unique_prices=('Price', 'nunique'))
    .query('unique_prices > 1')
)

print(f'Комбинаций с разными ценами: {len(price_check)}')

display(price_check.head())

Комбинаций с разными ценами: 1033


unique_prices
TransactionNo ProductNo               
536544        21218                  2
              22844                  2
              22862                  2
              22866                  2
536592        22165                  2

####**Рузультаты проверки и дальнейшие действия**

 В ходе проверки обнаружены две комбинации `TransactionNo` и `ProductNo`, в которых один и тот же товар имеет разные значения `Price`.

Объединение  по `TransactionNo` и `ProductNo` может привести к потере информации, это стоит учитывать
и поле `Price` включается в группировку.


---

 Отрицательные значения `Quantity `на данный момент не изменяются.

 Отрицательные значения `Quantity`  участвуют в объединении наравне с остальными записями, поскольку отражают отдельный тип операций (например, возвраты) и будут исследоваться далее отдельно.


 ---

Для сохранения данных,  позиции будут объединены по  `Quantity.`

Группировка выполняется по:

- `TransactionNo`
- `Date`
- `ProductNo`
- `ProductName`
- `Price`
- `CustomerNo`
- `Country`

In [58]:
# Объединение неявных дубликатов

df = (
    df.groupby(
        [
            'TransactionNo',
            'Date',
            'ProductNo',
            'ProductName',
            'Price',
            'CustomerNo',
            'Country'
        ],
        as_index=False
    )['Quantity']
    .sum()
)

print(f'Размер датафрейма после объединения: {df.shape}')

display(df.head())

Размер датафрейма после объединения: (526417, 8)


,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
0,536365,2018-12-01,21730,Glass Star Frosted T-Light Holder,14.61,17850,United Kingdom,6
1,536365,2018-12-01,22752,Set 7 Babushka Nesting Boxes,18.09,17850,United Kingdom,2
2,536365,2018-12-01,71053,White Moroccan Metal Lantern,13.72,17850,United Kingdom,6
3,536365,2018-12-01,84029E,Red Woolly Hottie White Heart,13.72,17850,United Kingdom,6
4,536365,2018-12-01,84029G,Knitted Union Flag Hot Water Bottle,13.72,17850,United Kingdom,6


In [59]:
remaining_duplicates = (
    df.groupby(['TransactionNo', 'ProductNo', 'Price'])
      .size()
      .gt(1)
      .sum()
)

print(f'Оставшихся неявных дубликатов: {remaining_duplicates}')

Оставшихся неявных дубликатов: 0


####**Итоги обработки**

- Неявные дубликаты успешно объединены.

- Повторяющиеся позиции одного товара внутри одной транзакции сведены в одну запись путем суммирования `Quantity.`

- Использование `Price` в ключе группировки позволило избежать объединения записей с различной стоимостью товара.

- Проверка показала, что после обработки неявные дубликаты отсутствуют.

- Датафрейм готов к дальнейшему исследованию.

///// *Общий вывыод по шагу 5*

-----
----
---

##**Шаг 6. Детальный анализ особенностей данных**

Выявление аномалий, особенностей распределений и скрытых закономерностей, которые могут повлиять на дальнейший анализ.

В рамках данного шага будут рассмотрены:
- распределение `Price`
-  распределение `Quantity` (включая отрицательные значения)
- исследование выбросов и длинных хвостов  
- анализ уникальных значений  
- поиск дополнительных закономерностей и аномалий



###**Анализ распределение `Price`**


Ранее при первом знакомстве с признаком `Price` были изучены общие статистические характеристики.

На данном шаге, подробнее рассмотрим особенности распределения цен, а также проверим природу анамальных высоких значений.




In [60]:
df['Price'].quantile([
    0.25,
    0.50,
    0.75,
    0.90,
    0.95,
    0.99
])

,Price
0.25,11.02
0.50,11.94
0.75,14.09
0.90,17.90
0.95,19.61
0.99,27.30


Большинство товаров имеет стоимость менее 27.30, однако для оценки наличия длинного правого хвоста рассмотрим количество наблюдений, превышающих 99-й перцентиль.

In [61]:
q99 = df['Price'].quantile(0.99)

print(df[df['Price'] > q99].shape[0])

4469


Полученные результаты свидетельствуют о наличии относительно небольшой группы товаров с существенно более высокой стоимостью. Для понимания природы этих наблюдений рассмотрим наиболее дорогие товары.

In [62]:
top_expensive_products = (
    df[['ProductName', 'Price']]
    .drop_duplicates()
    .sort_values(by='Price', ascending=False)
    .head(15)
)

top_expensive_products

,ProductName,Price
27486,Vintage Blue Kitchen Cabinet,660.62
4744,Vintage Red Kitchen Cabinet,660.62
80360,Vintage Red Kitchen Cabinet,594.50
130068,Vintage Blue Kitchen Cabinet,594.50
105952,Love Seat Antique White Metal,210.12
23930,Love Seat Antique White Metal,189.62
238,Rustic Seventeen Drawer Sideboard,179.37
18504,Regency Mirror With Shutters,179.37
18503,Rustic Seventeen Drawer Sideboard,158.88
83700,Regency Mirror With Shutters,158.88


Среди наиболее дорогих товаров встречаются одинаковые наименования с различными значениями цены. На данном этапе это рассматривается как особенность данных. Причины различий (изменение цен во времени, особенности продаж или другие факторы) будут исследованы на последующих этапах анализа.

####**Краткий вывод**

Анализ показал, что распределение Price имеет выраженный длинный правый хвост. Несмотря на то что основная часть товаров имеет стоимость ниже 99-го перцентиля (27.30), в данных присутствует небольшое количество значительно более дорогих товаров.

На текущем этапе признаков, позволяющих считать эти значения ошибками, не обнаружено, поэтому они сохраняются для дальнейшего анализа.

----
---
---

###**Анализ отрицательных  и аномальных значений Quantity**

- исследование структуры отклонений в количестве товаров;
- интерпретация отрицательных значений как возвратов;
- оценка объёма и доли аномалий;
- анализ распределения по транзакциям;
- принятие решения по обработке.






####**Отрицательное значение Quantity**

In [63]:
# Просмотр первых записей с отрицательным количеством
display( df[df["Quantity"] < 0] .sort_values(by="Quantity") .head(15) )

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
526411,C581484,2019-12-09,23843,Paper Craft Little Birdie,6.19,16446,United Kingdom,-80995
518850,C541433,2019-01-18,23166,Medium Ceramic Top Storage Jar,11.32,12346,United Kingdom,-74215
517930,C536757,2018-12-02,84347,Rotating Silver Angels T-Light Hldr,10.28,15838,United Kingdom,-9360
520587,C550456,2019-04-18,21108,Fairy Cake Flannel Assorted Colour,12.40,15749,United Kingdom,-3114
520588,C550456,2019-04-18,21175,Gin And Tonic Diet Metal Sign,12.15,15749,United Kingdom,-2000
520591,C550456,2019-04-18,85123A,Cream Hanging Heart T-Light Holder,12.86,15749,United Kingdom,-1930
521571,C556522,2019-06-13,22920,Herb Marker Basil,10.81,16938,United Kingdom,-1515
520589,C550456,2019-04-18,47566B,Tea Time Party Bunting,12.86,15749,United Kingdom,-1300
524452,C570556,2019-10-11,20971,Pink Blue Felt Craft Trinket Box,11.34,16029,United Kingdom,-1296
524014,C569552,2019-10-04,15034,Paper Pocket Traveling Fan,10.32,14533,United Kingdom,-1200


In [64]:
# Количество и доля отрицательных значений

negative_quantity_count = df[df["Quantity"] < 0].shape[0]
negative_quantity_share = negative_quantity_count / df.shape[0] * 100

print(f"Количество записей: {negative_quantity_count}")
print(f"Доля записей: {negative_quantity_share:.2f}%")

Количество записей: 8519
Доля записей: 1.62%


In [65]:

#Проверка записей  начинаются транзакций с "C"

(
  df.loc[df["Quantity"] < 0, "TransactionNo"]
  .astype("string")
  .str.startswith("C", na=False)
  .value_counts()
  )

,count
TransactionNo,
True,8519


In [66]:
#Поиск отрицательных значений без префикса "C"

negative_without_c = df[
    (df["Quantity"] < 0)
    & (
        df["TransactionNo"]
        .astype("string")
        .str.startswith("C", na=False) == False
    )
]

print(
    "Количество отрицательных записей без префикса C:",
    negative_without_c.shape[0]
)



Количество отрицательных записей без префикса C: 0


In [67]:
# Сохранение возвратов в отдельный датафрейм
returns_df = df[df["Quantity"] < 0].copy()

#Cоздание датафрейма с состоявшимися продажами
sales_df = df[df["Quantity"] > 0].copy()

----

#### **Аномальные значения Quantity**


In [68]:
#Создание датафрейма с положительными значениями Quantity

positive_quantity_df = df[df["Quantity"] > 0].copy()

print(
    "Количество записей с положительным Quantity:",
    positive_quantity_df.shape[0]
)

Количество записей с положительным Quantity: 517898


In [69]:
#Квантили положительных значений
sales_df["Quantity"].quantile([
    0.25,
    0.50,
    0.75,
    0.90,
    0.95,
    0.99
])

,Quantity
0.25,1.0
0.50,4.0
0.75,12.0
0.90,24.0
0.95,30.0
0.99,100.0


In [70]:
# Определение 99-го перцентиля
quantity_q99 = positive_quantity_df["Quantity"].quantile(0.99)

print(f"99-й перцентиль Quantity: {quantity_q99}")

99-й перцентиль Quantity: 100.0


In [71]:
#Выделение наблюдений выше 99-го перцентиля

high_quantity_df = positive_quantity_df[
    positive_quantity_df["Quantity"] > quantity_q99
].copy()

print(
    "Количество наблюдений выше 99-го перцентиля:",
    high_quantity_df.shape[0]
)

Количество наблюдений выше 99-го перцентиля: 4852


In [72]:
# Просмотр наблюдений с наибольшим количеством товара

largest_quantity_rows = (
    positive_quantity_df
    .sort_values(
        by="Quantity",
        ascending=False
    )
    .head(15)
)

display(largest_quantity_rows)

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
516441,581483,2019-12-09,23843,Paper Craft Little Birdie,12.38,16446,United Kingdom,80995
58955,541431,2019-01-18,23166,Medium Ceramic Top Storage Jar,11.32,12346,United Kingdom,74215
479617,578841,2019-11-25,84826,Asstd Design 3d Paper Stickers,6.19,13256,United Kingdom,12540
403126,573008,2019-10-27,84077,World War 2 Gliders Asstd Designs,10.47,12901,United Kingdom,4800
197065,554868,2019-05-27,22197,Popcorn Holder,10.99,13135,United Kingdom,4300
92989,544612,2019-02-22,22053,Empire Design Rosette,11.09,18087,United Kingdom,3906
259099,560599,2019-07-19,18007,Essential Balm 35g Tin In Envelope,10.31,14609,United Kingdom,3186
50323,540815,2019-01-11,21108,Fairy Cake Flannel Assorted Colour,12.40,15749,United Kingdom,3114
153278,550461,2019-04-18,21108,Fairy Cake Flannel Assorted Colour,12.40,15749,United Kingdom,3114
414686,573995,2019-11-02,16014,Small Chinese Style Scissor,10.58,16308,United Kingdom,3000


In [73]:
# Получение номеров соответствующих транзакций

largest_transaction_numbers = (
    largest_quantity_rows["TransactionNo"]
    .dropna()
    .unique()
)

display(largest_transaction_numbers)

array(['581483', '541431', '578841', '573008', '554868', '544612',
       '560599', '540815', '550461', '573995', '562439', '536830',
       '554272', '543057', '543669'], dtype=object)

In [74]:
#Просмотр всех позиций найденных транзакций

largest_quantity_transactions = (
    positive_quantity_df[
        positive_quantity_df["TransactionNo"].isin(
            largest_transaction_numbers
        )
    ]
    .sort_values(
        by=["TransactionNo", "Quantity"],
        ascending=[True, False]
    )
)

display(largest_quantity_transactions)

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
4701,536830,2018-12-02,84077,World War 2 Gliders Asstd Designs,10.43,16754,United Kingdom,2880
4700,536830,2018-12-02,21915,Red Harmonica In Box,11.34,16754,United Kingdom,1400
50323,540815,2019-01-11,21108,Fairy Cake Flannel Assorted Colour,12.40,15749,United Kingdom,3114
50324,540815,2019-01-11,21175,Gin And Tonic Diet Metal Sign,12.15,15749,United Kingdom,2000
50325,540815,2019-01-11,85123A,Cream Hanging Heart T-Light Holder,12.86,15749,United Kingdom,1930
58955,541431,2019-01-18,23166,Medium Ceramic Top Storage Jar,11.32,12346,United Kingdom,74215
77098,543057,2019-02-03,84077,World War 2 Gliders Asstd Designs,10.47,16333,United Kingdom,2592
77097,543057,2019-02-03,21915,Red Harmonica In Box,11.34,16333,United Kingdom,420
83689,543669,2019-02-11,22693,Grow A Flytrap Or Sunflower In Tin,11.21,16029,United Kingdom,2400
92989,544612,2019-02-22,22053,Empire Design Rosette,11.09,18087,United Kingdom,3906


#### **Итоги наблюдений**


- В ходе анализа установлено, что отрицательные значения `Quantity` составляют **1,62%** от общего числа наблюдений. Все такие записи относятся к транзакциям с префиксом `C`, что подтверждает их связь с возвратами или отменёнными операциями. Для последующего анализа продаж данные записи выделены в отдельный датафрейм.

- Анализ положительных значений показал, что **99-й перцентиль** распределения `Quantity` составляет **100 единиц**, а выше этого значения находится **4 852 наблюдения**. Просмотр крупнейших значений и соответствующих транзакций не выявил очевидных признаков ошибок: большинство экстремальных значений являются частью реальных крупных заказов.

- На данном этапе основания для удаления высоких значений `Quantity` отсутствуют. Возвраты исключены из анализа продаж, а положительные экстремальные значения сохранены для дальнейшего исследования.

----
----
---

### **Осмотр уникальных позиций**

- анализ количества уникальных значений в полях с категориальными данными;
- проверка корректности и согласованности значений;
- выявление редких или нестандартных позиций, требующих внимания.

In [75]:
# Быстрый взгляд на все категориальные поля
df[['Country',
    'CustomerNo',
    'ProductName',
    'TransactionNo']].nunique().sort_values(ascending=False)


,0
TransactionNo,23204
CustomerNo,4739
ProductName,3768
Country,38


### **Анализ хвостов и подсчет хваостов**

- нужно подсчитать что выше 99 квартиля и какая сумма формируется от общего число и понять выброс это или нет

In [76]:
df['Quantity'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])

,Quantity
0.25,1.0
0.50,3.0
0.75,11.0
0.90,24.0
0.95,30.0
0.99,100.0


---
---
---

#**//Общий вывод по первому этапу**

----
----
----

#**Этап 2. Предварительный анализ и создание новых признаков**

- Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
- Создание признаков: месяц, день недели, год из даты.
- Создание признака: общая сумма заказа (`Price` и `Quantity`).
- Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

### **Создание новых признаков из даты**

Для дальнейшего анализа сезонности и временных трендов, из колонки `Date` будут извлечены год, месяц и день недели.

In [77]:
# Извлечение года, месяца и дня недели
df['Year'] = df['Date'].dt.year
df['Month'] = df['Date'].dt.month
df['DayOfWeek'] = df['Date'].dt.dayofweek # Понедельник=0, Воскресенье=6

display(df.head())

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity,Year,Month,DayOfWeek
0,536365,2018-12-01,21730,Glass Star Frosted T-Light Holder,14.61,17850,United Kingdom,6,2018,12,5
1,536365,2018-12-01,22752,Set 7 Babushka Nesting Boxes,18.09,17850,United Kingdom,2,2018,12,5
2,536365,2018-12-01,71053,White Moroccan Metal Lantern,13.72,17850,United Kingdom,6,2018,12,5
3,536365,2018-12-01,84029E,Red Woolly Hottie White Heart,13.72,17850,United Kingdom,6,2018,12,5
4,536365,2018-12-01,84029G,Knitted Union Flag Hot Water Bottle,13.72,17850,United Kingdom,6,2018,12,5


### **Агрегация данных на уровне заказа для анализа корзины**

Для анализа содержимого корзины и среднего чека, агрегируем данные по `TransactionNo`, чтобы получить общую сумму и количество уникальных товаров в каждом заказе.

### **Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен**

Давайте исследуем, какие товары являются наиболее прибыльными и популярными, а также как распределены цены на товары.

### **Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме**

Проанализируем структуру заказов, чтобы понять типичный размер корзины, средний чек и как распределяются заказы по их общей стоимости.

# **Этап 3. Исследовательский анализ данных (EDA)**

*   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
*   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
*   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
*   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
*   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
*   Анализ ценовых сегментов: зависимость количества продаж от цены.

# **Этап 4. Ключевые выводы и рекомендации**
*   Формирование основных инсайтов по каждому блоку анализа.
*   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.

----
----
----